Sc 9790 questionnaires adr (#5242)
simpledotorg/simple-server: doc/arch/020-questionnaires-aka-dynamic-forms.md

Static forms have served us well for infrequently changing forms like Blood Pressure and Blood Sugar. For forms that require frequent changes (like Drug-stock), we used WebView to accommodate the dynamic nature of this form. WebView isn't offline-first, like the rest of Simple app.

Questionnaires (aka Dynamic forms) is an effort at solving 2 challenges in an offline-first manner: reducing time to production & a flexibility to add/remove fields without need to update the Simple app. We also refer to them as Server-driven UI.

## Decision

We have designed Questionnaires by reusing underlying offline-first sync architecture of Simple. Definitions of new terminology introduced:
- **Form:** A document to gather information in a structured manner.
- **Questionnaire:** A sync resource containing form's layout and input fields, version and form-type.
- **Questionnaire Response:** A sync resource containing user's inputs for a form/questionnaire, and user/facility the input was recorded at.

> - **Layout:** A custom DSL that helps Mobile rendering a form defined by server.

Author: @tfidfwastaken incorporating this suggestion as it's a bit verbose & clearer.

> ## Design (AKA Implementation Details)

Author: done.

Questionnaires are implemented by following below mentioned steps:

### 1. Mobile _syncs_ questionnaires from server to device
Mobile requests questionnaires from server using [the questionnaire sync API](https://api.simple.org/api-docs/index.html#tag/Questionnaires). Server responds with all layouts compatible with the DSL version supported by the mobile app.

> On Mobile database, the primary key for Questionnaires sync resource is questionnaire_type, and not id. _That is a key difference between Questionnaire and other sync resources._ Reason being, mobile needs only 1 layout per type of a questionnaire. Whenever a new questionnaire of an existing type is created, server sends the updated ID & layout in the sync response, and mobile updates the row for corresponding type.

Author: > How is this possible if `questionnaire_type` is the primary key? Wouldn't the above be an update then?

Added server-side for clarity `new questionnaire of an existing type is created on Server-side`.

> On Mobile database, the primary key for Questionnaires sync resource is questionnaire_type, and not id. _That is a key difference between Questionnaire and other sync resources._ Reason being, mobile needs only 1 layout per type of a questionnaire. Whenever a new questionnaire of an existing type is created, server sends the updated ID & layout in the sync response, and mobile updates the row for corresponding type.

Author: Accepted style suggestion.

> On Mobile database, the primary key for Questionnaires sync resource is questionnaire_type, and not id. _That is a key difference between Questionnaire and other sync resources._ Reason being, mobile needs only 1 layout per type of a questionnaire. Whenever a new questionnaire of an existing type is created, server sends the updated ID & layout in the sync response, and mobile updates the row for corresponding type.

Author: > If an event is unique or its date or time is known, use when. Whenever is best used for repeated events or events whose date or time is uncertain.

I'm using whenever to communicate that this time is uncertain.

> created vs updated
Questionnaires are meant to be immutale. We will never modify an existing questionnaire, rather create a new one & mark older as inactive.

| questionnaire_type <PK> | id     | layout   |
|-------------------------|--------|----------|
| screening_reports       | uuid-1 | {form-1} |
| supplies_reports        | uuid-2 | {form-2} |

### 2. Mobile _syncs_ questionnaire_responses from server to device

> User inputs for a dynamic form are stored in a questionnaire response. Mobile requests them from server using [this sync API](https://api.simple.org/api-docs/index.html#tag/Questionnaire-Responses).

Author: This is for mobile syncing from server. Section # 6 & 7 talks about mobile sending questionnaires to server.

> User inputs for a dynamic form are stored in a questionnaire response. Mobile requests them from server using [this sync API](https://api.simple.org/api-docs/index.html#tag/Questionnaire-Responses).

Reviewer: Yeah. It's unclear because when you read that user inputs are stores in a response, you expect the server to fetch the response.

```suggestion
User inputs for a dynamic form are stored in a questionnaire response. Mobile requests response objects from the server using [this sync API](https://api.simple.org/api-docs/index.html#tag/Questionnaire-Responses) so they are available to the mobile user to store inputs.
```

> User inputs for a dynamic form are stored in a questionnaire response. Mobile requests them from server using [this sync API](https://api.simple.org/api-docs/index.html#tag/Questionnaire-Responses).

Author: Understood. Reworded the sentence to use questionnaire response sync resource (instead of response objects).

Other sync resources have explicit keys for recording input (Blood Pressures have "systolic"/"diastolic", Patients have "gender", etc.). Questionnaire response records all user inputs in a single field `content` of JSON data type.

| id      | questionnaire_id | questionnaire_type | facility_id | user_id | content          |
|---------|------------------|--------------------|-------------|---------|------------------|

> | uuid-11 | uuid-1           | screening_reports  | Surat       | 123     | {"name": "john"} |

Author: done

| uuid-22 | uuid-2           | drug_stock_reports | Bangalore   | 456     | {"name": "doe"}  |

### 3. Displaying questionnaires on home page
![questionnaires-on-home-page](resources/questionnaires-home-page.png)

> The decision to display or hide link to a dynamic form on the home page is based on 3 steps:

Reviewer 2: word choice suggestion:
```suggestion
The decision to display or hide link to a dynamic form on the home page is based on 3 criteria:
```

> The decision to display or hide link to a dynamic form on the home page is based on 3 steps:

Author: All of them must be met. Have updated the doc to reflect this nuance.

1. Mobile has a questionnaire of given type.
1. Mobile has questionnaire responses for given type.

> 1. Config of a questionnaire is enabled for User's facility, which mobile derives from [the facility sync API](https://api.simple.org/api-docs/index.html#tag/facility).

Reviewer 2: clarity suggestion:
```suggestion
1. Config for the questionnaire type is enabled for User's facility, which mobile derives from [the facility sync API](https://api.simple.org/api-docs/index.html#tag/facility).
```

> 1. Config of a questionnaire is enabled for User's facility, which mobile derives from [the facility sync API](https://api.simple.org/api-docs/index.html#tag/facility).

Reviewer: If config is just a flipper flag, can say it.

Author: This is not a flipper flag, but a config field in the facility table. Updated the doc to mention that.

### 4. Displaying a list of forms
![questionnaire-responses-list](resources/questionnaire-responses-list.png)

When user clicks on a form type from home page, Mobile queries its table for responses of given questionnaire type. Mobile has some custom logic on ordering and style of the forms:

> 1. Mobile orders forms in descending order of months by checking for `content["month_date"]` key.

Reviewer: What happens if mobile can't find a response for a month_date?

Author: > If Mobile cannot find these keys in a questionnaire response, it will crash.

Added this line to ADR.

1. Mobile deduces submission status by checking for `content["submitted"]` key.
1. Mobile localizes month_date before displaying them.

> 1. This implies that `month_date` & `submitted` keys are _"mandatory/static"_ in monthly questionnaire responses.

Reviewer 2: Question: What does `mandatory/static` mean here? This is the first time this terminology shows up, maybe it would be helpful to clarify what that means?

Author: Reworded this. Does it make sense now?
> To maintain backward/forward compatibility of a response when new questionnaire layouts are released (more details in FAQ section below), all keys inside a content are meant to be treated as optional. However, for monthly reports, `month_date` & `submitted` keys are _mandatory_ for mobile app to display them.

### 6. User submits a form
When user submits a form, Mobile updates the `content`, `questionnaire_id`(to record the layout which was displayed to a user), `last_updated_by_user_id` & `facility_id` for a questionnaire response.

A questionnaire's layout contains a `link_id` for input fields. Mobile uses them to display values in UI and record responses in the content key. 
A specimen content key's value looks like:
```json
{
  "submitted": true,
  "month_date": "08-2023",
  "monthly_screening_report.total_bp_checks.male": 10,
  "monthly_screening_report.comments": "All good"
}
```

### 7. Handling concurrent updates
When server receives different questionnaire_responses for the same ID, it merges the 2 records in the following way:
1. Latest `device_updated_at` takes precedence for every field, except `content`.
1. A UNION of existing & new `content` is taken.
1. For conflicting keys within content, latest update takes precedence.
1. Consider this specimen workflow:
    1. User #1 fills keys A, B, C & D on 6 Dec
    1. User #2 fills keys C, D, E & F on 7 Dec
    1. Post union, DB will contain keys A, B, C, D, E & F
    1. C, D will be User #2’s data.

### 8. User changes locale

> For most screens, mobile stores the translations for all languages. Since a questionnaire's layout is driven by server, translations are also stored on server-side. When user changes a locale, a sync gets triggerd on mobile. We leveraged this behaviour of mobile to communicate layouts with new translations from server to mobile. 

Reviewer 2: typo fix and prose suggestion:
```suggestion
For most screens, mobile stores the translations for all languages. Since a questionnaire's layout is driven by server, translations are also stored on server-side. When user changes a locale, a sync gets triggered on mobile. We leverage this behaviour of mobile to communicate layouts with new translations from server to mobile. 
```

> We're storing locale inside Questionnaire sync resource's `process_token` (_another aspect where questionnaires differ from other sync resources_). When a user changes locale, Server **resyncs** all questionnaires to Mobile because of mismatch between user & process token's locale.

Reviewer 2: Having difficulty understanding the last sentence. Did you mean:
> mismatch between [user locale] and [process token locale](?)
What are the consequences of these two being mismatched?

Author: Having an understanding of how sync resources work is a prerequisite for this ADR. Basically, process_token is a cursor to mark how many resources have been synced to device & which to sync next. It can define which resources to sync next, or can signal a full resync based on mismatches like region change (or locale change in the case of questionnaires)

Reviewer 2: I am familiar with sync resources and I've read the ADR already :)

I looked at the questionnaire code and got my answer. My confusion was around the phrase `user and process token mismatch`. It is ambiguous whether `user` refers to `user locale`, `token sent by user` or something else.

I think I would have more easily understood that point if the last line were written somewhat like this:

> In addition to the usual keys, we're storing the user's locale inside Questionnaire sync resource's `process_token`. The resync logic in questionnaires is different from other resources, in that it triggers a full resync of all questionnaires when the user's locale changes. This is to ensure that all the questionnaires will have the correct translations when the locale changes.

Author: Done. Updated the language.

> ### 9. FAQs

Reviewer 2: Can call this section "Other features and points of note", since there's no question being asked in this FAQ
